benchmarks/distributed/rpc/rl/agent.py

```diff
@@ -0,0 +1,165 @@
+from functools import reduce
+import time
+import threading
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+
+import torch.distributed.rpc as rpc
+from torch.distributed.rpc import RRef, remote
+from torch.distributions import Categorical
+
+from torch import autograd
+
+OBSERVER_NAME = "observer{}"
+gamma = 0.99
+
+torch.autograd.set_detect_anomaly(True)
+
+
+class Policy(nn.Module):
+    def __init__(self, state_size, nlayers, out_features, batch=True):
+        torch.autograd.set_detect_anomaly(True)
+        super(Policy, self).__init__()
+        self.in_features = reduce((lambda x, y: x*y), state_size)
+
+        self.model = nn.Sequential(
+            nn.Flatten(1, -1),
+            nn.Linear(self.in_features, out_features),
+            * [nn.Linear(out_features, out_features) for _ in range(nlayers)]
+        )
+        self.dim = 0
+
+    def forward(self, x):
+        action_scores = self.model(x)
+        return F.softmax(action_scores, dim=self.dim)
+
+
+class AgentBase:
+    def __init__(self, batch=True):
+        self.id = rpc.get_worker_info().id
+        self.running_reward = 0
+        self.eps = 1e-7
+
+        self.ob_rrefs = []   # Observer RRef
+        self.rewards = {}
+
+        self.future_actions = torch.futures.Future()
+        self.lock = threading.Lock()
+
+        self.agent_latency_start = None
+        self.agent_latency_end = None
+        self.agent_latency = []
+        self.agent_throughput = []
+
+    def reset_metrics(self):
+        self.agent_latency_start = None
+        self.agent_latency_end = None
+        self.agent_latency = []
+        self.agent_throughput = []
+
+    def set_world(self, batch_size, state_size, nlayers, out_features, batch=True):
+        from observer import ObserverBase
+
+        self.batch = batch
+        self.policy = Policy(state_size, nlayers, out_features, self.batch)
+        self.optimizer = optim.Adam(self.policy.parameters(), lr=1e-2)
+
+        self.batch_size = batch_size
+        for rank in range(batch_size):
+            ob_info = rpc.get_worker_info(OBSERVER_NAME.format(rank + 2))
+            self.ob_rrefs.append(
+                remote(ob_info, ObserverBase))
```

Comment:
This will create one set of observers. But the coordinator also calls `ob_ref = remote(ob_info, ObserverBase)`, there there will be two different set of observers. We only need one of them. Does it make sense to let the coordinator create the observers and pass their RRefs to the agent?